queue/claimresolver.js

```diff
@@ -0,0 +1,243 @@
+var Promise       = require('promise');
+var debug         = require('debug')('queue:claimresolver');
+var slugid        = require('slugid');
+var assert        = require('assert');
+var _             = require('lodash');
+var base          = require('taskcluster-base');
+var data          = require('./data');
+var QueueService  = require('./queueservice');
+
+/** State that are considered resolved */
+const RESOLVED_STATES = [
+  'completed',
+  'failed',
+  'exception'
+];
+
+/**
+ * Facade that handles resolution of claims by takenUntil, using the advisory
+ * messages from the azure queue. The azure queue messages takes the form:
+ * `{taskId, runId, takenUntil}`, and they become visible after `takenUntil`
+ * has been exceeded. The messages advice that if a task with the given
+ * `takenUntil` and `taskId` exists, then `runId` maybe need to be resolved by
+ * `claim-expired` and the task retried (depending on `retriesLeft`).
+ *
+ * Notice, that the task may not exists, or may not have a the given run, or
+ * it may be the case that the `takenUntil` doesn't match. These should be
+ * ignored as it implies a failed request (which wasn't retried) or a run
+ * that was reclaimed.
+ */
+class ClaimResolver {
+  /**
+   * Create ClaimResolver instance.
+   *
+   * options:
+   * {
+   *   Task:           // instance of data.Task
+   *   queueService:   // instance of QueueService
+   *   publisher:      // publisher from base.Exchanges
+   *   pollingDelay:   // Number of ms to sleep between polling
+   *   parallelism:    // Number of polling loops to run in parallel
+   *                   // Each handles up to 32 messages in parallel
+   * }
+   */
+  constructor(options) {
+    assert(options, "options must be given");
+    assert(options.Task.prototype instanceof data.Task,
+           "Expected data.Task instance");
+    assert(options.queueService instanceof QueueService,
+           "Expected instance of QueueService");
+    assert(options.publisher, "Expected a publisher");
+    assert(typeof(options.pollingDelay) === 'number',
+           "Expected pollingDelay to be a number");
+    assert(typeof(options.parallelism) === 'number',
+           "Expected parallelism to be a number");
+    this.Task         = options.Task;
+    this.queueService = options.queueService;
+    this.publisher    = options.publisher;
+    this.pollingDelay = options.pollingDelay;
+    this.parallelism  = options.parallelism;
+
+    // Promise that polling is done
+    this.done         = null;
+    // Boolean that polling should stop
+    this.stopping     = false;
+  }
+
+  /** Start polling */
+  start() {
+    if (this.done) {
+      return;
+    }
+    this.stopping = false;
+
+    // Start a loop for the amount of parallelism desired
+    var loops = [];
+    for(var i = 0; i < this.parallelism; i++) {
+      loops.push(this.poll());
+    }
+    // Create promise that we're done looping
+    this.done = Promise.all(loops).catch((err) => {
```

Comment:
A better option would be have to producer/consumer pattern... where the producer pulls from azure queue and consumer handles x number of messages in parallel.

Because the issue here is that we want to run x handle of messages in parallel... This is quick hack until I figure out a good pattern for how to do this nicely...